extensions/binding/org.eclipse.smarthome.binding.mqttbroker/src/main/java/org/eclipse/smarthome/binding/mqttbroker/internal/discovery/MqttServiceDiscoveryService.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.binding.mqttbroker.internal.discovery;
+
+import java.util.*;
+
+import org.eclipse.smarthome.binding.mqttbroker.MqttBrokerBindingConstants;
+import org.eclipse.smarthome.config.discovery.*;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.io.transport.mqtt.*;
+import org.osgi.service.component.annotations.*;
+import org.slf4j.*;
+
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link MqttServiceDiscoveryService} is responsible for discovering connections on
+ * the MqttService shared connection pool.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@Component(immediate = true, service = DiscoveryService.class, name = "MqttServiceDiscoveryService")
+public class MqttServiceDiscoveryService extends AbstractDiscoveryService implements MqttBrokersObserver {
+    private final Logger logger = LoggerFactory.getLogger(MqttServiceDiscoveryService.class);
+    MqttService mqttService;
+
+    public MqttServiceDiscoveryService() {
+        super(Sets.newHashSet(MqttBrokerBindingConstants.THING_TYPE_TEXTUAL,
+                MqttBrokerBindingConstants.THING_TYPE_BROKER), 500, true);
+    }
+
+    @Override
+    @Activate
+    protected void activate(Map<String, Object> config) {
+        super.activate(config);
+    };
+
+    @Override
+    @Deactivate
+    protected void deactivate() {
+        super.deactivate();
+    }
+
+    @Reference
+    public void setMqttService(MqttService service) {
+        mqttService = service;
+        if (isBackgroundDiscoveryEnabled()) {
+            startBackgroundDiscovery();
+        }
+    }
+
+    public void unsetMqttService(MqttService service) {
+        mqttService = null;
+        if (isBackgroundDiscoveryEnabled()) {

```

Comment:
You must not do that after your component is deactivated. Move this to deactivate.

Reply:
Hm, but unsetMqttService can only be called while the component is active. What does the OSGi specification say about referenced components and the unbind method?

Reply:
No, it is only called when it is NOT active.